crates/binjs_generate_library/src/lib.rs

```diff
@@ -429,13 +386,49 @@ impl<'a> Walker<'a> for {name} where Self: 'a {{
                     let definition = format!(
                         "
 /// Implementation of interface sum {node_name}
-#[derive(PartialEq, Debug, Clone)]
+#[derive(PartialEq, Debug, Clone, Serialize)]
+#[serde(tag = \"type\")]
 pub enum {name} {{
 {contents}
     /// An additional value used to mark that the node was stolen by a call to `steal()`.
     BinASTStolen,
 }}\n
 
+// An optimised implementation of tagged deserialise that expects `type` to be the first key in the object.
+//
+// This does not strictly adhere to JSON spec, but gives ~2.5x better performance than generic
+// deserialistaion with arbitrary ordering.
+//
+// See https://github.com/serde-rs/serde/issues/1495 for details.
+impl<'de> serde::Deserialize<'de> for {name} {{
+    fn deserialize<D: serde::Deserializer<'de>>(de: D) -> Result<Self, D::Error> {{
+        #[derive(Deserialize)]
+        enum VariantTag {{
+            {variant_tags}
+        }}
+
+        struct MapVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for MapVisitor {{
+            type Value = {name};
+
+            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {{
```

Comment:
I don't see where this function is used.

Reply:
Also, what exactly *is* the error message if the type doesn't come first?

Reply:
It's implementation of Serde trait. It's not used directly by us.

Reply:
Just checked and tried to pass invalid JSON:

> `Could not parse source: JSONError(Error("unknown variant `expression`, expected `type`", line: 1, column: 3205177))`

I guess it could be better, I just was lazy and used a simple `enum` with a single variant.

What do you think - should I implement custom deserialize there to provide a better message?

Reply:
If it's not complicated, yeah, that would be nicer.

Because the day we notice this error message, I'm sure that I will have forgotten what it means :)

Reply:
Yeah, agreed.

Reply:
Updated to:

> Could not parse source: JSONError(Error("invalid value: key `expression`, expected a `type` key as the first item in the object", line: 1, column: 3205177))

Reply:
Much better, thanks!

Perhaps emphasize word `first`?

Reply:
Like how?

Reply:
Like `*first*` or `FIRST` :)